### ISSUE TITLE
ParamSpec-based signature change works with functions but not classes

### ISSUE THREAD

**Issue by brentyi** [microsoft/pyright#4813]
Hi!

I'm running into an issue where a `ParamSpec`-based signature works for function inputs, but breaks for a class that has the exact same signature.

**Background**

I have an application with a method `f()` that has an extremely vague signature:
```python
class Class:
    def f(self, *args, **kwargs) -> None:
        print(args, kwargs)
```

And I want to write a decorator that populates the signature of `f` by borrowing from another function (evokes functools.wraps):

```python
def dummy_f(a: str, b: float) -> None:
    ...

class Class:
    @replace_signature_with(func_with_desired_signature)
    def f(self, *args, **kwargs) -> None:
        print(args, kwargs)
```

Since `f()` is a method, I wrote a decorator that concatenates `Class` as the first argument to a ParamSpec. So `dummy_f`'s signature `(a: str, b: float)` gets transformed into `(self: Class, a: str, b: float)`:

```python
def replace_signature_with(
    func_with_desired_signature: Callable[P, Any]
) -> Callable[[Callable], Callable[Concatenate[Class, P], None]]:
    """Decorator for replacing the signature of a function with that of some other function."""
    return lambda f: f  # type: ignore
```

Putting everything together, this works great!

```python
from __future__ import annotations

import dataclasses
from typing import Any, Callable, Concatenate, ParamSpec, cast

P = ParamSpec("P")


def replace_signature_with(
    func_with_desired_signature: Callable[P, Any]
) -> Callable[[Callable], Callable[Concatenate[Class, P], None]]:
    """Decorator for replacing the signature of a function with that of some other function."""
    return lambda f: f  # type: ignore


def dummy_f(a: str, b: float) -> None:
    ...


class Class:
    @replace_signature_with(dummy_f)
    def f(self, *args, **kwargs) -> None:
        print(args, kwargs)


x = Class()
reveal_type(x.f)  # Type of "x.f" is "(a: str, b: float) -> None"
```

**The problem**

What seems to break, however, is when I swap the `dummy_f` _function_ with a `Dummy` _class_:

```python
from __future__ import annotations

import dataclasses
from typing import Any, Callable, Concatenate, ParamSpec, cast

P = ParamSpec("P")


def replace_signature_with(
    func_with_desired_signature: Callable[P, Any]
) -> Callable[[Callable], Callable[Concatenate[Class, P], None]]:
    """Decorator for replacing the signature of a function with that of some other function."""
    return lambda f: f  # type: ignore


@dataclasses.dataclass
class Dummy:  # New!
    a: str
    b: float


class Class:
    @replace_signature_with(Dummy)  # dummy_f => Dummy
    def f(self, *args, **kwargs) -> None:
        print(args, kwargs)


x = Class()
reveal_type(x.f)  # Type of "x.f" is "(Class, a: str, b: float) -> None"
## ^ This looks wrong!!
```

`Dummy` and `dummy_f` from earlier have the same signature, but for some reason `Class.f` seems to be now interpreted as a static method, where the first argument is no longer as interpreted as `self` and dropped.

Is this a bug?

Thanks!!

**Comment by erictraut** [comment]
I agree that there is an inconsistency here, and I consider that a bug. The question is what the correct (consistent) behavior should be. I need to think about that further.

The challenge here is that a `Callable` in the Python type system does not indicate whether the resulting callable should be treated as an instance method, a class method, or a static method (or standard function). That is, when a Callable is accessed through a class instances, should it be "bound" to that instance (and a `self` argument automatically provided to the call)? Generally, pyright assumes that `Callable` means a static method, in which case no binding occurs. However, when the `Callable` includes a ParamSpec, the signature captured by the ParamSpec dictates whether the final Callable should be treated as an instance method, class method, or static method. 

Mypy appears to be using some different heuristic to determine whether to treat a `Callable` in a class as an instance method or static method. I'll need to investigate further to understand its behavior here and whether it makes sense.

**Comment by erictraut** [comment]
This will be fixed in the next release. The fix turned out to be relatively simple. When a class is used in place where a function is expected, it is assumed that the class constructor should be used. Pyright synthesizes this function from the class's `__init__` and/or `__new__` methods. When it did so, it set an internal flag indicating that the function was a static method. This flag was captured and propagated by the ParamSpec, so the resulting function returned by the decorator also had this flag set. The fix is to clear this flag.

**Comment by brentyi** [comment]
Makes sense, thanks. :)

One bit I'm wondering about is static methods. With the fix, it sounds like the behavior of the snippet I sent earlier will match what happens if we pass in a regular function:
```python
@dataclasses.dataclass
class Dummy:
    a: str
    b: float

class Class:
    @replace_signature_with(Dummy)
    def f(self, *args, **kwargs) -> None:
        print(args, kwargs)

x = Class()
reveal_type(x.f)  # (after fix) Type of "x.f" is "(a: str, b: float) -> None"
```

What if we pass in a static method? Will the following result in different behavior?
```python
@dataclasses.dataclass
class Dummy:
    a: str
    b: float

    @staticmethod
    def make(a: str, b: float) -> Dummy:  # New!
        return Dummy(a, b)


class Class:
    @replace_signature_with(Dummy.make)
    def f(self, *args, **kwargs) -> None:
        print(args, kwargs)
```

**Comment by erictraut** [comment]
In this case, the flag that indicates it's a static method will be captured by the ParamSpec, and the resulting decorated function `f` will also be considered a static method by the type checker. That may not be what you intend here, but I think it's the right behavior for the type checker. You'd need to use a different approach in that case.

**Comment by brentyi** [comment]
Got it. Do you have any examples off the top of your head where propagating the static method flag is used / useful?

I checked in mypy and it behaves differently, where `Dummy` (the constructor) and `Dummy.make` (the static method) are handled identically when used in a ParamSpec-annotated decorator and no static flag seems to be captured. As an equivalent example, while these two are equivalent in Pyright:
```python
class Class:
    @replace_signature_with(Dummy.make)
    def f(self, *args, **kwargs) -> None:
        print(args, kwargs)
```
```python
class Class:
    @staticmethod  # Added
    @replace_signature_with(Dummy.make)
    def f(self, *args, **kwargs) -> None:
        print(args, kwargs)
```
In mypy, only the second example treats `Class.f` as a static method. From my narrow point of view this seems more natural/useful to me compared to the signature that pyright produces, which implies that I'm calling `staticmethod()` inside of the decorator.

**Comment by erictraut** [comment]
Here's an example of where it's important to copy the internal flags as part of the `ParamSpec` capture.

```python
def my_decorator(func: Callable[P, R]) -> Callable[P, R]:
    return func

class Class:
    @my_decorator
    @staticmethod
    def f(a: int, b: str) -> None:
        ...
```

**Comment by brentyi** [comment]
> > Here's an example of where it's important to copy the internal flags as part of the ParamSpec capture.
>
> This makes perfect sense as a very useful feature.

1. In terms of practical usefulness: for cases like the identity decorator example it doesn't seem like you need ParamSpec at all; just a TypeVar + `bound=Callable` would do, no?

2. In general: I suppose my position is more based on the fact that I've never seen a decorator wrapped around `@staticmethod` in practice, likely because `staticmethod()` doesn't return actual callables in Python <=3.9. All of the standard decorators, `@functools.lru_cache`, `@abc.abstractmethod`, etc need to be inside of the `@staticmethod` decorator, and there isn't much you can do with a decorator that wraps around `@staticmethod` beyond the toy identity decorator example provided. I see now in Python 3.10 that `staticmethod()` does now return a callable with a valid signature... frustrating.

--- 

@erictraut to address my issue without affecting the identity decorator example, if we have:
```python
class Class:
    @staticmethod
    def f(a: int, b: str) -> None:
        ...
```
would it be possible to clear the "static method" flag when we access the function via `Class.f`? While `staticmethod()` returns a wrapped version of `f`, note that this gets unwrapped when we access it via `Class.f`:

```python
class Class:
    @staticmethod
    def f(a: int, b: str) -> None:
        ...


print(isinstance(staticmethod(lambda a, b: None), staticmethod))  # Prints `True`
print(isinstance(Class.f, staticmethod))  # Prints `False`
```
This would leave the behavior of your example untouched, enable the `@replace_signature_with()` decorator I had described, and also help match Pyright's behavior closer to mypy.

**Comment by erictraut** [comment]
I'm going to leave pyright's current behavior in place for now. This is under-specified in PEP 612. Pyright has used this same implementation for the past two+ years (and has had it in place much longer than mypy). Changing it is likely to create churn for existing pyright users. Unless there's a really compelling reason to change it, I think it's best to leave it as is.

**Comment by brentyi** [comment]
Sure, I think use cases for the proposed change are rare and on my end switching away from `ParamSpec` is not too much effort. I appreciate the discussion and can just do that. 🙂 

For posterity / in case you decide to revisit stripping that static flag when we access `Class.some_static_method`, here's an example with a fuller motivation for this issue. I'm trying to use a decorator factory to help instantiate messages and send them along a websocket connection. It would be nice one day to be able to pass static methods into the decorator factory:

```python
from __future__ import annotations

import dataclasses
import functools
from typing import Any, Callable
from typing_extensions import Concatenate, ParamSpec

P = ParamSpec("P")


def _message_sender(
    message_cls: Callable[P, Any]
) -> Callable[[Callable], Callable[Concatenate[MessageServer, P], None]]:
    """We don't want to re-write every single message type definition; this decorator
    factory instead helps us (a) instantiate a message and (b) use the server API
    to send it."""

    def inner(self: MessageServer, *args: P.args, **kwargs: P.kwargs) -> None:
        message = message_cls(*args, **kwargs)
        self._send(message)

    return lambda _: inner


@dataclasses.dataclass
class MessageA:
    a: int


@dataclasses.dataclass
class MessageB:
    b: int

    @staticmethod
    def make(b: int) -> MessageB:
        return MessageB(b)


class MessageServer:
    def _send(self, message: Any) -> None:
        ...

    @_message_sender(MessageA)
    def send_a(self):
        ...

    @_message_sender(MessageB.make)
    def send_b(self):
        ...


if __name__ == "__main__":
    server = MessageServer()
    server.send_a(a=5)  # Mypy OK, Pyright OK in next version.
    server.send_b(b=5)  # Mypy OK, but Pyright errors.

```

**Comment by erictraut** [comment]
This is included in pyright 1.1.301, which I just published. It will also be included in a future release of pylance.
